aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -67,6 +80,16 @@ using namespace at::sparse_csr;
 // certain utiliy functions are usable from sparse COO.
 using namespace at::sparse;
 
+namespace {
+
+    inline Tensor get_result_tensor_for_unary_op(const Tensor& input) {
+        if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
```

Comment:
It doesn't work without this if-statement?

Reply:
Thanks for the question, Ivan. No (I just checked), without this if-statement, it will raise the following error: (for inputs with integral types)

```python
RuntimeError: result type Float can't be cast to the desired output type Long
```

The reason is, unary ops promote to default type for integral inputs.

Reply:
Isn't that something we'd also support?

Reply:
Thanks for the question, @cpuhrsch. I think yes, I had a discussion offline with @IvanYashchuk about this. Here is how different variants of `torch.sin` act w.r.t to an integral tensor as the input:

```python
In [2]: x = torch.tensor([1, 2, 3, 4], dtype=torch.int64)

In [3]: torch.sin(x)
Out[3]: tensor([ 0.8415,  0.9093,  0.1411, -0.7568])

In [4]: x.sin_()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-91f1e9cc2c47> in <module>
----> 1 x.sin_()

RuntimeError: result type Float can't be cast to the desired output type Long

In [5]: torch.sin(x, out=x.clone())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-cbf65c5f71e1> in <module>
----> 1 torch.sin(x, out=x.clone())

RuntimeError: result type Float can't be cast to the desired output type Long
```

The interesting case is with `torch.sin`, and for consistency - we will also want to support `torch.sin(input)` for Sparse CSR tensors. Even though the old implementation from this PR works (please see below), there is a better version suggested by @IvanYashchuk which is to calculate the result from the op's variant first and then create an output sparse CSR tensor from itself with it's dtype.

<details>
<summary>(Old) Code</summary>

```cpp
  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
      return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
    }
    return at::empty_like(input);
  }
```
</details>

<details>
<summary>(New) Code</summary>

```cpp
  template <typename F>
  inline Tensor get_result_tensor_for_unary_op(F op, const Tensor& input) {
    auto values = input.values();

    // To handle type promotion for inputs to unary ops,
    // we first get the result from the underlined op, and use the result
    // to create a sparse CSR tensor, which is used as the input to the out= variant
    auto result_values = (*op)(values);

    auto result = at::native::sparse_csr_tensor(
      input.crow_indices().clone(),
      input.col_indices().clone(),
      result_values,
      result_values.scalar_type(),
      input.layout(),
      result_values.device());

    return result;
  }
```
</details>

This also required a fix as done here: https://github.com/pytorch/pytorch/pull/68123/files#diff-550964753dcb07747fac5a9559da33e00603c96a99c4f6eafb0359b38722680bR277-R300. There can be a cleaner version, which I'll look at (to just calculate the dtype in the if-else condition, and use it further).